Feature/improve parser

taylorhansen opened the pull request:
This adds a number of improvements to the current grammar, fixing a couple bugs with the AST generation and merging all the grammar files into one 431-line "megaparser".

I made this branch off of `feature/class-fields` a while ago so it's showing the commits made for that branch too. Just pay attention to 4e72247 and everything beyond that.

### I have:
 - [x] Ran a test (`npm run dev && npm test`)
 - [x] Added tests for my changes
 - [x] Documented my code thoroughly.

taylorhansen committed:
Fix statements documentation

They used to be Node[]'s, but now it's wrapped around a CodeBlock.

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
Add protected access modifiers

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
Add newline to end of file

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
Fix type being undefined

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
parser: Add fields to classes

Also added some tests to make sure they're parsing.

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
grammar: Merge all grammars into one "megaparser"

Like the name? It's only 431 lines anyways so you don't have to keep a
bunch of tabs open.

This also gets rid of a nasty circular dependency between parser.ne and
function.ne because both need a production rule from eachother. An ugly
hack to resolve this was by using a template rule.

ImmutableDictionary was removed because pretty useless anyways,
considering the fact that the production rule is already commented out.
I'm not sure what was being thought there with adding an "immutable"
dictionary to begin with.

The 4 literals Array, Dictionary, Tuple, and Set had Node constructors
that take in some VSLTokenType thing which isn't really needed and it
broke those current production rules, so the type argument was removed
from their Node constructors.

CommandChains took up a bunch of space in the grammar and, since it's
commented out, it should be removed. We have version control anyways so
we can restore it if we really need it.

A short PatternDeclaration production rule had the same fate here, and
TypeAlias wasn't referenced anywhere else so that was removed, too.

Lastly, some production rules were rearranged to make the grammar
easier to read.

taylorhansen committed:
Merge branch 'develop' into feature/class-fields

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
Move parser.ne out of its own folder

That grammar file really needs to get out every once in a while.

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
Merge branch 'develop' into feature/improve-parser

vihanb commented on src/vsl/parser/nodes/typeAlias.js:
```diff
@@ -1,27 +0,0 @@
-import Node from './node';
```
why remove this?

taylorhansen replied:
I found the production rule for `TypeAlias` but it wasn't referenced anywhere so I just deleted it.

vihanb replied:
hm, we were going to use them as a variation of variable declarations:

```swift
public typealias GoatBarn = Barn<Goat>
```

maybe leave it and we can finish up in future PR? We could also just re-add it in the future too though since it's all tracked

taylorhansen replied:
Let's re-add it later. It's better to let version control handle this stuff.

vihanb commented on src/vsl/parser/parser.ne:
```diff
@@ -0,0 +1,430 @@
+
+@builtin "postprocessors.ne"
+
+CodeBlock[s]
```
This is kinda difficult to read being so long so I think adding some:

```ruby
# ================== Data Structures ================
```

would help give a sense of hierarchy in the parser. Though @somebody1234 has worked on parser most and I don't want this to be non-compatible with his branch >_<

taylorhansen replied:
I should probably do that, yea.

somebody1234 replied:
I'm fine with manually merging

taylorhansen committed:
readability: Rearrange some things

Also added section comments to provide some better structure to the
grammar file.

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
fix: Allow no arguments in a function call

Fixes #31.

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

taylorhansen committed:
test: Add tests for function calls

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>

vihanb commented on src/vsl/parser/parser.ne:
```diff
@@ -0,0 +1,458 @@
+
+FunctionCallList
+   -> delimited[FunctionCallArgument {% id %}, _ "," _]:? {%
+        (data, location) => new t.FunctionCall(null, data[0], location)
```
Since you have `:?` I believe it should be `data[0] || []` as `data[0]` can be null whereas it should always be an array

taylorhansen replied:
I forgot about that. :P

somebody1234 reviewed (commented):
Looks fine, I'm fine with merging after you add tests

taylorhansen committed:
Fix undefined function call list

Signed-off-by: CrazyGuy108 <taylor@hansenfamily.us>
